File: spk/API/Entities/UserGoal.cs
namespace API.Entities;

public class UserGoal : BaseEntity
{
    public string UserId { get; set; } = string.Empty;

    public AppUser? User { get; set; }

    public string Title { get; set; } = string.Empty;

    public string? Description { get; set; }

    public UserGoalType GoalType { get; set; }

    public int TargetValue { get; set; }

    public int CurrentValue { get; set; }

    public DateTime StartDate { get; set; } = DateTime.UtcNow;

    public DateTime TargetDate { get; set; } = DateTime.UtcNow.AddDays(30);

    public UserGoalStatus Status { get; set; } = UserGoalStatus.Active;

    public DateTime? CompletedAt { get; set; }
}

File: Client/src/models/userGoal.ts
export const UserGoalType = {
  QuestionCount: 1,
  StudyMinutes: 2,
  CompletedTopics: 3,
  CompletedCourses: 4,
  TrialExamCount: 5,
  ReviewCount: 6,
} as const

export type UserGoalType = (typeof UserGoalType)[keyof typeof UserGoalType]

export const UserGoalStatus = {
  Active: 1,
  Completed: 2,
  Archived: 3,
} as const

export type UserGoalStatus = (typeof UserGoalStatus)[keyof typeof UserGoalStatus]

export interface UserGoal {
  id: string
  title: string
  description?: string | null
  goalType: UserGoalType
  targetValue: number
  currentValue: number
  progressPercentage: number
  startDate: string
  targetDate: string
  daysRemaining: number
  status: UserGoalStatus
  isOverdue: boolean
  completedAt?: string | null
  createdAt: string
}

export interface CreateUserGoalRequest {
  title: string
  description?: string | null
  goalType: UserGoalType
  targetValue: number
  startDate: string
  targetDate: string
}

export interface UpdateUserGoalRequest {
  title: string
  description?: string | null
  targetValue: number
  startDate: string
  targetDate: string
  status: UserGoalStatus
}
